lib/bundlex/toolchain/common/unix/os_deps.ex

```
@@ -0,0 +1,175 @@
+    os_deps =
+      parse_os_deps(native.os_deps)
+      |> Enum.map(fn
+        {:pkg_config, lib_names} ->
+          {:pkg_config, lib_names}
+
+        {precompiled_dependency, lib_names} ->
+          case maybe_download_precompiled_package(precompiled_dependency) do
+            :unavailable ->
+              # fallback
+              {:pkg_config, lib_names}
+
+            package_path ->
+              {{precompiled_dependency, package_path}, lib_names}
+          end
+      end)
+
+    %{native | os_deps: os_deps}
+  end
+
+  defp maybe_download_precompiled_package(precompiled_dependency_url) do
+    File.mkdir_p(@precompiled_path)
+    package_path = get_package_path(precompiled_dependency_url)
+
+    cond do
+      package_path == :unavailable ->
+        :unavailable
+
+      File.exists?(package_path) ->
+        package_path
+
+      true ->
+        try do
+          File.mkdir(package_path)
+          temporary_destination = "#{@precompiled_path}/temporary"
+          download(precompiled_dependency_url, temporary_destination)
+          System.shell("tar -xf #{temporary_destination} -C #{package_path} --strip-components 1")
+          System.shell("rm #{temporary_destination}")
+          package_path
+        rescue
+          e ->
+            Logger.warning("Couldn't download the dependency due to: #{inspect(e)}.")
+            :unavailable
+        end
+    end
+  end
+
+  defp get_package_path(:unavailable), do: :unavailable
+
+  defp get_package_path(url) do
+    url = if String.ends_with?(url, "/"), do: String.slice(url, 0..-2), else: url
+
+    last_part =
+      String.split(url, "/")
+      |> Enum.at(-1)
+      |> String.split(".")
+      |> Enum.reject(&(&1 in ["tar", "xz", "gz"]))
+      |> Enum.join(".")
+
+    "#{@precompiled_path}#{last_part}"
+  end
+
+  defp download(url, dest) do
+    response = Req.get!(url)
+
+    if response.status == 200 do
+      File.write(dest, response.body)
+    else
+      :error
+    end
+  end
```

Value returned from this function is discarded. Take it into account, or raise instead of returning `:error`

---

done